File: Server/Server-CSharp/YT/Helper/MsgHelper.cs
using ENet;
using Google.Protobuf;
using System;
using System.IO;
using System.Linq;

namespace YT
{
	class MsgHelper
	{
	
		/// <summary>
		/// 编码
		/// </summary>
		/// <param name="msg"></param>
		/// <returns></returns>
		public static byte[] Encode(IMessage msg) {
			//数据编码
			byte[] nameBytes = EncodeName(msg);
			byte[] bodyBytes = EncodeBody(msg);
            int len = nameBytes.Length + bodyBytes.Length;
            byte[] sendBytes = new byte[2 + len];
            //组装长度
            sendBytes[0] = (byte)(len % 256);
            sendBytes[1] = (byte)(len / 256);
            //组装名字
            Array.Copy(nameBytes, 0, sendBytes, 2, nameBytes.Length);
            //组装消息体
            Array.Copy(bodyBytes, 0, sendBytes, 2 + nameBytes.Length, bodyBytes.Length);
            return sendBytes;
        }

		/// <summary>
		/// 将protobuf对象序列化位byte数组
		/// </summary>
		/// <param name="msg"></param>
		/// <returns></returns>
		private static byte[] EncodeBody(IMessage msg) {
			return msg.ToByteArray();
			
		}

		/// <summary>
		/// 解码
		/// </summary>
		/// <returns></returns>
		public static Request Decode(ByteArray readBuff,Peer peer) {
            //消息长度
            if (readBuff.length <= 2) {
                return null;
            }
            //消息体长度
            int readIdx = readBuff.readIdx;
            byte[] bytes = readBuff.bytes;
            Int16 bodyLength = (Int16)((bytes[readIdx + 1] << 8) | bytes[readIdx]);
            if (readBuff.length < bodyLength) {
                return null;
            }
            readBuff.readIdx += 2;
            //解析协议名
            int nameCount = 0;
            string protoName = MsgHelper.DecodeName(readBuff.bytes, readBuff.readIdx, out nameCount);
            if (protoName == "") {
                Console.WriteLine("OnReceiveData MsgBase.DecodeName fail");
                //Close(user);
                return null;
            }
            readBuff.readIdx += nameCount;
            //解析协议体
            int bodyCount = bodyLength - nameCount;
            if (bodyCount < 0) {
                Console.WriteLine("OnReceiveData fail, bodyCount <0 ");
                //Close(user);
                return null;
            }
			byte[] msg = DecodeBody( readBuff.bytes, readBuff.readIdx, bodyCount);
            readBuff.readIdx += bodyCount;
            readBuff.CheckAndMoveBytes();
            Request request = new Request();
            request.Peer = peer;
            request.Name = protoName;
            request.Msg = msg;
            return request;
        }
	    
		//解码协议体
	    private static byte[] DecodeBody( byte[] bytes, int offset, int count) {
			byte[] bodyBytes = new byte[count];
			Array.Copy(bytes, offset, bodyBytes, 0, count);
	        return bodyBytes;
	    }
	    
		//编码协议名（2字节长度+字符串）
	    private static byte[] EncodeName(IMessage msg){
			//名字bytes和长度
			byte[] nameBytes = System.Text.Encoding.UTF8.GetBytes(msg.GetType().Name);
	        short len = (short)nameBytes.Length;
			//申请bytes数值
			byte[] bytes = new byte[2+len];
			//组装2字节的长度信息
			bytes[0] = (byte)(len%256);
			bytes[1] = (byte)(len/256);
			//组装名字bytes
			Array.Copy(nameBytes, 0, bytes, 2, len);
	
			return bytes;
		}
	
		//解码协议名（2字节长度+字符串）
		private static string DecodeName(byte[] bytes, int offset, out int count){
			count = 0;
			//必须大于2字节
			if(offset + 2 > bytes.Length){
				return "";
			}
	        //读取长度
	        int len = ((bytes[offset+1] << 8 )| bytes[offset] );
			if(len <= 0){
				return "";
			}
			//长度必须足够
			if(offset + 2 + len > bytes.Length){
				return "";
			}
			//解析
			count = 2+len;
			string name = System.Text.Encoding.UTF8.GetString(bytes, offset+2, len);
			return name;
		}
	}
	
}

File: Server-Go/YT/Helper/MsgHelper.go
package Helper

import (
	"ReaminHistory/YT/Base"
	"fmt"
	"github.com/golang/protobuf/proto"
	"google.golang.org/protobuf/runtime/protoiface"
	"reflect"
)

//编码
func Encode(msg protoiface.MessageV1) []byte {
	nameBytes := encodeName(msg)
	bodyBytes := encodeBody(msg)
	//组装长度
	length := len(nameBytes) + len(bodyBytes)
	sendBytes := make([]byte, 2)
	sendBytes[0] = byte(length % 256)
	sendBytes[1] = byte(length / 256)
	//组装协议名
	sendBytes = append(sendBytes, nameBytes...)
	//组装协议体
	sendBytes = append(sendBytes, bodyBytes...)
	return sendBytes
}

func encodeBody(msg protoiface.MessageV1) []byte {
	buffer, _ := proto.Marshal(msg)
	return buffer
}

func encodeName(msg protoiface.MessageV1) []byte {
	name := reflect.TypeOf(msg).String()
	sendBytes := make([]byte, 2)
	sendBytes[0] = byte((len(name) - 7) % 256)
	sendBytes[1] = byte((len(name) - 7) / 256)
	sendBytes = append(sendBytes, []byte(name)[7:]...)
	return sendBytes
}

func Decode(readBuf *Base.ByteArray, conv uint32) *Base.Request {
	bytes := readBuf.Read()
	fmt.Println(len(bytes), "  ", string(bytes))
	//解析协议名
	protoName, nameCount := decodeName(bytes)
	if protoName == "" {
		fmt.Println("OnReceiveData MsgBase.DecodeName fail")
		return &Base.Request{}
	}
	//解析协议体
	if len(bytes) < nameCount {
		fmt.Println("OnReceiveData fail, bodyCount <0")
		return &Base.Request{}
	}
	bytes = bytes[nameCount:]
	msg := decodeBody(bytes)
	request := &Base.Request{Conv: conv, Name: protoName, Msg: msg}
	return request
}

func decodeName(bytes []byte) (name string, count int) {
	//必须大于两字节
	if len(bytes) < 2 {
		return "", 0
	}
	//读取长度
	length := int16(bytes[1])*256 + int16(bytes[0])

	if length < 0 {
		return "", 0
	}
	//长度必须足够
	if int(length)+2 > len(bytes) {
		return "", 0
	}
	name = string(bytes[2 : length+2])
	return name, int(length + 2)
}

func decodeBody(bytes []byte) []byte {
	return bytes
}
